File: C++ Programs/Sorts/quickSort.cpp
#include<iostream>
using namespace std;

int split(int arr[], int lower, int upper)
{
    int pivot, low, high;

    low = lower + 1;
    high = upper;
    pivot = arr[lower];

    while(high >= low)
    {
        while(arr[low] < pivot)
        {
            low++;
        }
        while(arr[high] > pivot)
        {
            high--;
        }
        if(high > low)
        {
            int temp = arr[low];
            arr[low] = arr[high];
            arr[high] = temp;
        }
    }

    int temp = arr[lower];
    arr[lower] = arr[high];
    arr[high] = temp;
    return high;
}

void quickSort(int arr[], int lower, int upper)
{
    int pivot;
    if(upper > lower)
    {
        pivot = split(arr, lower, upper);
        quickSort(arr, lower, pivot - 1);
        quickSort(arr, pivot + 1, upper);
    }
}

int main()
{
    int arr[] = {11, 2, 9, 13, 57, 25, 17, 1, 90, 3, 99};
    cout<<"Quick Sort\nBefore Sorting : ";
    for(int i = 0; i < 9 ; i++)
    {
        cout<<arr[i]<<" ";
    }
    quickSort(arr, 0, 9);
    cout<<"\nAfter Sorting : ";
    for(int i = 0; i < 9 ; i++)
    {
        cout<<arr[i]<<" ";
    }
    return 0;
}

File: Java Programs/sort/quicksort.java
//QUICK SORT program
public class quicksort
{
  
public static void main(String args[] )
{
	int arr[] = { 11, 2, 9, 13, 57, 25, 17, 1, 90, 3, 99 } ;
	int i ;

	System.out.print ( "Quick sort.\n" ) ;
	System.out.print ( "\nArray before sorting:\n") ;

	for ( i = 0 ; i <= 9 ; i++ )
		System.out.print( "\t"+ arr[i] ) ;

	quicksort ( arr, 0, 9 ) ;

	System.out.print ( "\nArray after sorting:\n") ;

	for ( i = 0 ; i <= 9 ; i++ )
		System.out.print ( "\t"+ arr[i] ) ;

}
//Quick sort Recursive Function
public static void quicksort ( int a[ ], int lower, int upper )
{
	int i ;
	if ( upper > lower )
	{
		i = split ( a, lower, upper ) ;
		quicksort ( a, lower, i - 1 ) ;
		quicksort ( a, i + 1, upper ) ;
	}
}
//Function to divide array into two parts
public static int split ( int a[ ], int lower, int upper )
{
	int i, p, q, t ;

	p = lower + 1 ;
	q = upper ;
	i = a[lower] ;

	while ( q >= p )
	{
		while ( a[p] < i )
			p++ ;

		while ( a[q] > i )
			q-- ;

		if ( q > p )
		{
			t = a[p] ;
			a[p] = a[q] ;
			a[q] = t ;
		}
	}
	t = a[lower] ;
	a[lower] = a[q] ;
	a[q] = t ;
	return q ;
}
}
